cdap-watchdog/src/main/java/io/cdap/cdap/logging/ErrorLogsClassifier.java

```diff
@@ -219,6 +255,56 @@ private void populateResponse(IThrowableProxy throwableProxy,
     }
   }
 
+  @Nullable
+  private ErrorClassificationResponse findMatchingRule(IThrowableProxy throwableProxy,
+      ErrorClassificationResponse ruleMatchedResponsePrev) {
+    ErrorClassificationResponse ruleMatchedResponse = null;
+    for (ErrorClassificationRule rule : ruleList) {
+      if (classNameMatched(rule.getExceptionClassRegex(), throwableProxy.getClassName())
+          && codeMethodMatched(rule.getCodeMethodRegex(),
+          throwableProxy.getStackTraceElementProxyArray())) {
+
+        String errorMessage = throwableProxy.getMessage();
+        ruleMatchedResponse = new ErrorClassificationResponse.Builder()
+            .setErrorCategory(ErrorCategoryEnum.OTHERS.name())
+            .setErrorType(rule.getErrorType().name())
+            .setErrorMessage(errorMessage)
+            .setDependency(String.valueOf(rule.isDependency()))
+            .setThrowableClassName(throwableProxy.getClassName())
+            .setRuleId(rule.getId())
+            .setRulePriority(rule.getPriority())
+            .build();
+        break;
+      }
+      throwableProxy = throwableProxy.getCause();
+    }
+    if (ruleMatchedResponse == null) {
+      return ruleMatchedResponsePrev;
+    } else if (ruleMatchedResponsePrev == null) {
+      return ruleMatchedResponse;
+    } else {
+      return ruleMatchedResponse.getRulePriority() > ruleMatchedResponsePrev.getRulePriority() ?
+          ruleMatchedResponse : ruleMatchedResponsePrev;
+    }
+  }
+
+  private boolean codeMethodMatched(Pattern codeMethodRegex,
+      StackTraceElementProxy[] stackTraceElementProxies) {
+    if (codeMethodRegex == null) {
```

Comment:
consider doing field `Predicate<String> codeMethodPredicate`, set it to `ste -> codeMethodRegex.matcher(ste.toString()).find()` and set it to `TRUE::booleanValue` or similar by default. This would remove need for this check and make next expression simpler. Ditto for other cases

Reply:
realized this `null` check is not needed. Removed it.